src/Compilers/CSharp/Portable/Symbols/Source/ConstantEvaluationHelpers.cs

```diff
@@ -233,148 +234,95 @@ private struct Node<T> where T : class
+            ref ArrayBuilder<SourceFieldSymbolWithSyntaxReference> fieldsInvolvedInCycles)
         {
             Debug.Assert(graph.Count > 0);
 
-            var compilations = PooledDictionary<Compilation, int>.GetInstance();
-            OrderCompilations(graph, compilations);
-            var comparer = new SourceLocationComparer(compilations);
-
-            // Sort the fields by lexical order.
-            var fields = ArrayBuilder<SourceFieldSymbolWithSyntaxReference>.GetInstance();
-            fields.AddRange(graph.Keys);
-            fields.Sort(comparer);
-
-            var field = GetMemberOfCycle(graph);
-            fields.Clear();
-
-            GetAllReachable(graph, fields, field);
-            fields.Sort(comparer);
-
-            // Return the first field lexically in the cycle.
-            field = fields[0];
-            fields.Free();
-            compilations.Free();
-            return field;
-        }
-
-        /// <summary>
-        /// Return an ordering of the compilations referenced in the graph.
-        /// The actual ordering is not important, but we need some ordering
-        /// to compare source locations across different compilations.
-        /// </summary>
-        private static void OrderCompilations(
-            Dictionary<SourceFieldSymbolWithSyntaxReference, Node<SourceFieldSymbolWithSyntaxReference>> fields,
-            Dictionary<Compilation, int> compilations)
-        {
-            Debug.Assert(fields.Count > 0);
-            Debug.Assert(compilations.Count == 0);
-
-            // Note this order will not be consistent across multiple threads
-            // that are evaluating different graphs that start at a different
-            // field but overlap. But that should not matter since we need
-            // to sort fields within a cycle consistently across threads but
-            // not necessarily across cycles, and all fields in a cycle must
-            // be within the same compilation.
-            foreach (var field in fields.Keys)
+            if (fieldsInvolvedInCycles is null)
             {
-                var compilation = field.DeclaringCompilation;
-                if (!compilations.ContainsKey(compilation))
-                {
-                    compilations.Add(compilation, compilations.Count);
-                }
+                fieldsInvolvedInCycles = ArrayBuilder<SourceFieldSymbolWithSyntaxReference>.GetInstance(graph.Count);
+                // We sort fields that belong to the same compilation by location to process cycles in deterministic order.
+                // Relative order between compilations is not important, cycles do not cross compilation boundaries. 
+                fieldsInvolvedInCycles.AddRange(graph.Keys.GroupBy(f => f.DeclaringCompilation).
+                    SelectMany(g => g.OrderByDescending((f1, f2) => g.Key.CompareSourceLocations(f1.ErrorLocation, f2.ErrorLocation))));
```

Comment:
Consider marking the outer lambda static. I assume making the inner one static will be extremely complicated as it's a comparer?